### PR TITLE
Fix S1854 FP: When a variable is used inside local function

### LINKED ISSUE #3126: Fix S1854 FP: When a variable is used inside local function
### Description

When a variable is used inside local function, S1854 is raised.

### Repro steps

        public string Test()
        {
            string buffer = new string('a', 1);
            return Local();

            string Local()
            {
                return buffer;
            }
        }

### Expected behavior

No warning.

### Actual behavior

warning S1854: Remove this useless assignment to local variable 'buffer'.

### Related information

* SonarAnalyzer.CSharp.8.2.0.14119
* VisualStudio 2019 16.4.4

### LINKED ISSUE #3719: Fix S1854 FP: Variable used in local function as delegate
Separate issue for case reported in https://github.com/SonarSource/sonar-dotnet/issues/3169#issuecomment-671863449 by @Rubbiroid.

```C#
public void StoreEvents(IList<IEvent> events)
{
    var debouncedEventTypes = new HashSet<Type>();
    var debouncedEvents = events.Where(NotYetDebounced);

    foreach (var debouncedEvent in debouncedEvents)
    {
        StoreEvent(debouncedEvent);
    }

     bool NotYetDebounced(IEvent @event)
     {
         return !(@event is IWithDebounce) || debouncedEventTypes.Add(@event.GetType());
     }
}
```

sonar says that debouncedEventTypes is not used: Remove this useless assignment to local variable S1854

We have Sonar Version 8.3.1 (build 34397) and C# plugin 8.6.1 (build 17183)

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3126
Fixes #3719

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
I don't want to fix the old coverage issue in this case.

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/MutedSyntaxWalker.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                SyntaxKind.MethodDeclaration,
+                SyntaxKind.PropertyDeclaration,
+                SyntaxKind.GetAccessorDeclaration,
+                SyntaxKind.SetAccessorDeclaration,
+                SyntaxKind.AddAccessorDeclaration,
+                SyntaxKind.RemoveAccessorDeclaration,
+                SyntaxKind.AnonymousMethodExpression,
+                SyntaxKind.SimpleLambdaExpression,
+                SyntaxKind.ParenthesizedLambdaExpression
+            };
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/MutedSyntaxWalker.cs:51]: NITPICK: indentation does not align with the opening curly bracket.

**Comment by csaba-sagi-sonarsource** [comment]
I know we said we will not cover formatting stuff now, but please fix `IsAllowedNumericInitialization` in DeadStores.cs the method is  unreadable.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**